packages/indexer-agent/src/network.ts

```diff
@@ -112,16 +124,123 @@ export class Network {
-      blockHash: receipt.blockHash,
+    const txPromise = transaction(Math.ceil(estimatedGas.toNumber()))
+    let tx = await txPromise
+
+    let txConfig: Record<string, number> = {
+      attempt: 1,
+      nonceOffset: 0,
+      gasBump: 1 + this.gasBumpPercent / 100,
+      nonce: tx.nonce,
+      gasPrice: tx.gasPrice.toNumber(),
+      gasLimit: tx.gasLimit.toNumber(),
+    }
+
+    logger.info(`Sending transaction`, { tx: tx, attempt: txConfig.attempt })
+
+    while (pending) {
+      if (txConfig.attempt >= this.maxRetries) {
+        logger.warn('Transaction attempt threshold surpassed, moving on', {
+          attempts: txConfig.attempt,
+        })
+        break
+      }
+
+      try {
+        if (txConfig.attempt > 1) {
+          logger.info('Resubmitting transaction', {
+            txConfig: txConfig,
+          })
+          const transaction: providers.TransactionRequest = {
+            value: tx.value,
+            to: tx.to,
+            data: tx.data,
+            chainId: tx.chainId,
+            from: tx.from,
+            nonce: txConfig.nonce,
+            gasPrice: Math.ceil(txConfig.gasPrice),
+            gasLimit: Math.ceil(txConfig.gasLimit),
+          }
+          tx = await this.wallet.sendTransaction(transaction)
+        }
+
+        logger.info(`Transaction pending`, { tx: tx })
+
+        const receipt = await this.ethereum.waitForTransaction(
+          tx?.hash,
+          3,
+          this.gasBumpTimeLimit,
+        )
+
+        if (receipt.status == 0) {
+          const tx = await this.ethereum.getTransaction(receipt.transactionHash)
+          const code = await this.ethereum.call(tx)
+          throw Error(`Transaction reverted, reason: ${code}`)
+        }
+
+        logger.info(`Transaction successfully included in block`, {
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          tx: tx?.hash,
+          receipt: receipt,
+        })
+        output = receipt
+        pending = false
+      } catch (error) {
+        txConfig = await this.updateTransactionConfig(logger, txConfig, error)
+        continue
+      }
+    }
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    return output!
+  }
+
+  async updateTransactionConfig(
+    logger: Logger,
+    txConfig: Record<string, number>,
+    error: Error,
+  ): Promise<Record<string, number>> {
+    if (
+      error.message.includes(
+        'Transaction with the same hash was already imported',
+      )
+    ) {
+      txConfig.nonceOffset += 1
+      txConfig.nonce =
+        (await this.wallet.getTransactionCount('pending')) -
+        txConfig.nonceOffset +
+        1
+    } else if (
+      error.message.includes(
+        'Transaction nonce is too low. Try incrementing the nonce.',
+      )
+    ) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      txConfig.nonce! += 1
+    } else if (error.message.includes('Reverted')) {
```

Comment:
This condition would also catch a valid revert and retry it. Let say that for some reason the agent is closing an already closed allocation, it will revert and then try again with higher gas limit and fail again.

Probably we can know if it was an out of gas revert by looking at the gasUsed being at least ~90% close to the gasLimit as we are adding a buffer of 1.5x for gas limit for the gas estimation.

Reply:
Good idea. I had been thinking that the contract calls beforehand were enough of a check for other revert reasons, but I like this additional protection.  I'll see if I can add a check for gas used being near the limit.

Reply:
Yes, the contracts calls we are doing for gas estimation are good because they will hit a high gas limit error if they will fail, however in the blockchain we can run into some time-sensitive conditions, for example:

1) Does gas estimation on an allocate() transaction and it won't revert.
2) Send a allocate() tx, takes some minutes to mine and when it gets mined some condition make it revert, like the indexer unstaked a chunk.
3) It will enter the revert retry logic.